### ISSUE TITLE
`__init__` takes precedence over `__new__`

### ISSUE THREAD

**Issue by ShineyDev** [microsoft/pyright#4569]
**Describe the bug**
Pyright appears to ignore type definitions in ``__new__`` if an ``__init__`` is defined.

**To Reproduce**
See below. Moving the ``__init__`` definition into an ``if not TYPE_CHECKING`` block produces the expected behavior.

**Expected behavior**
Typing in ``__new__`` should take precedence over ``__init__``.

**Screenshots or Code**
```python
class Foo:
    @overload
    def __new__(
        cls: type[Self],
        data: T_json_object,
    ) -> Self:
        ...

    @overload
    def __new__(
        cls: type[Self],
        data: list[T_json_object],
    ) -> list[Self]:
        ...

    def __new__(
        cls: type[Self],
        data: T_json_object | list[T_json_object],
    ) -> Self | list[Self]:
        if isinstance(data, list):
            return [cls(o, http) for o in data]

        return super().__new__(cls)

    def __init__(
        self: Self,
        data: T_json_object,
    ) -> None:
        self.data: T_json_object = data
```

**VS Code extension or command-line**
``pyright 1.1.292``

**Additional context**
N/A

**Comment by erictraut** [comment]
Pyright is working as designed here. It checks both the `__init__` and `__new__` for constructor calls. This is the standard behavior for constructors. What you're doing here is highly unusual.

While your code is legal Python, it won't work with static type checkers like pyright and mypy. Static type checkers need to make certain basic assumptions about the way Python code works. This is such an example.

If you don't want to change your code, you can use `# type: ignore` to suppress the error.
